torch/_functorch/functional_assertions_helper.py

```diff
@@ -0,0 +1,135 @@
+from typing import Callable, Dict, Optional
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+import torch._functorch.config as config
+
+aten = torch.ops.aten
+_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+
+
+class DepTokenStateTracker:
+    """
+    A singleton / global context manager which is used to:
+    - Record the global flag that assertion ops functionalization is enabled
+      during compilation if the context is entered.
+    - Record most recent token which is propagated through functional assertion
+      ops.
+    """
+
+    _dep_token = None
+    _entered: bool = False
+
+    def __enter__(self) -> "DepTokenStateTracker":
+        assert (
+            not DepTokenStateTracker._entered
+        ), "Cannot reenter singleton context `DepTokenStateTracker`."
+        self.reset()
+        DepTokenStateTracker._entered = True
+        return self
+
+    def __exit__(self, exc_type, exc_cal, exc_tb) -> None:
+        assert DepTokenStateTracker._entered, (
+            "Cannot exit singleton context `DepTokenStateTracker` which is not "
+            "entered previously."
+        )
+        self.reset()
+        DepTokenStateTracker._entered = False
+        return None
+
+    @classmethod
+    def reset(cls) -> None:
+        cls._dep_token = None
+
+    @classmethod
+    def set_dep_token(cls, dep_token) -> None:
+        assert cls._entered
+        cls._dep_token = dep_token
+
+    @classmethod
+    def get_dep_token(cls):
+        if cls._dep_token is None:
+            cls._dep_token = aten._make_dep_token()
+
+        return cls._dep_token
+
+
+class FunctionalAssertionsHelper:
+    @staticmethod
+    def can_functionalize_asserts(needs_autograd: bool) -> bool:
+        if config.functionalize_assertion_ops:
+            # The following assertions is aim to limit assertions functionalization
+            # to simple case for now.
+            # Only handle forward graph.
+            assert (
+                not needs_autograd
+            ), "Cannot functionalize assertion ops when grad is enabled"
+            # Avoid rng ops functionalization which also add extra outputs.
+            assert (
+                not config.functionalize_rng_ops
+            ), "Cannot functionalize assertion ops when RNG functionalization is enabled"
+
+        return config.functionalize_assertion_ops
+
+    @staticmethod
+    def functionalization_enabled() -> bool:
+        return DepTokenStateTracker._entered
+
+    @staticmethod
+    def get_decompositions() -> Dict[str, Dict[OpOverload, Callable]]:
+        return _decompositions
+
+    @staticmethod
+    def get_state_tracker() -> DepTokenStateTracker:
+        return DepTokenStateTracker()
+
+    @staticmethod
+    def create_compile_tracing_wrapper(func) -> Callable:
+        def _traced_forward(*args):
+            outs = func(*args)
+            return outs + (DepTokenStateTracker.get_dep_token(),)
```

Comment:
One more thing: right now, this means that if you set `config.functionalize_asserts=True`, then we will **always** end up putting a dependency token in the graph, even if there were not actually any asserts. This is potentially wasteful: we're allocating and returning an extra tensor in the graph for no reason.

Can we avoid this? One way would be if in this function, you check whether or not we've already created a dependency token: if not, that means that we don't need to return it here, and that we won't need the extra epilogue logic.

Reply:
Good catch. I originally was trying to avoid creating dep token if there is no assertion in the graph. However, this seems a little bit hard since I want to set `num_outputs_dep_token` in output of `run_functionalized_fw_and_collect_metadata` which is before the compiling? But let me double check on it, maybe I can record the whether there is assertion in the graph or not during `run_functionalized_fw_and_collect_metadata` since seems like we run the module anyway?